lib/ABI/FunctionType.cpp

```diff
@@ -388,210 +417,447 @@ class ConversionHelper {
+        return Used.begin()->Type;
       } else {
-        if constexpr (RegisterCount == 0)
-          return std::nullopt;
-        if (AllowedRegisters.front() == UsedRegisters.begin()->Location) {
-          if constexpr (DryRun)
-            return model::QualifiedType{};
-          else
-            return getTypeOrDefault(UsedRegisters.begin()->Type,
-                                    UsedRegisters.begin()->Location,
-                                    TheBinary);
-        } else {
-          return std::nullopt;
-        }
+        // Otherwise return a `void *` pointer as the return value type.
+        static constexpr auto VoidKind = model::PrimitiveTypeKind::Void;
+        auto Result = ensurePrimitiveTypeIsAvailable(VoidKind,
+                                                     0,
+                                                     Binary,
+                                                     TypeDependencies);
+        using model::Qualifier;
+        Result.Qualifiers.emplace_back(Qualifier::createPointer(Architecture));
+        return Result;
+      }
+    }
+
+    // TODO: there are some cases where preserving the total structure of
+    // the return value can be benefitial, this needs more work.
+    // For now, just determine the total size of the return value and if it
+    // fits into a primitive value - emit it, otherwise show those values as
+    // a struct (/note: positional ABIs need a whole lot of extra love in this
+    // regard).
+    llvm::SmallVector<std::optional<model::QualifiedType>, 4> ResultingTypes;
+    bool MustUseTheNextOne = false;
+    auto AllowedRange = llvm::enumerate(llvm::reverse(Allowed));
+    for (auto Pair : AllowedRange) {
+      size_t Index = Allowed.size() - Pair.index() - 1;
+      model::Register::Values Register = Pair.value();
+      auto UsedIterator = Used.find(Register);
+      if (UsedIterator != Used.end())
+        ResultingTypes.emplace_back(UsedIterator->Type);
+      else if (MustUseTheNextOne)
+        ResultingTypes.emplace_back(std::nullopt);
+
+      MustUseTheNextOne = MustUseTheNextOne || (UsedIterator != Used.end());
+    }
+
+    revng_assert(ResultingTypes.size() != 0);
+    if (ResultingTypes.size() == 1) {
+      // Short-circuit for a single register return value.
+      auto Kind = model::Register::primitiveKind(Allowed.at(0));
+      if (ResultingTypes.front() != std::nullopt) {
+        return ensureTypeIsAvailable(*ResultingTypes.front(),
+                                     Allowed.at(0),
+                                     Binary,
+                                     TypeDependencies);
+      } else {
+        // No information about the type of this register, use the default.
+        return ensureTypeIsAvailable(model::QualifiedType{},
+                                     Allowed.at(0),
+                                     Binary,
+                                     TypeDependencies);
```

Comment:
Why not push this directly in `ResultingTypes` instead of going through `std::nullopt`?

Reply:
`ensureTypeIsAvailable` took care of ensuring the primitive type actually exists (so it could fail). Since it's gone now (at least in my working branch), so is this issue.

Reply:
As mentioned in the previous comment, this is no longer relevant as the type manager now takes care of all this.